subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/modulecache/ModuleMetaDataCache.java

```diff
@@ -38,5 +40,17 @@
         boolean isMissing();
 
         ModuleSource getModuleSource();
+
+        /**
+         * The metadata after being processed by component metadata rules.
+         * Will be null the first time an entry is read from the filesystem cache during a build invocation.
+         */
+        @Nullable
+        ModuleComponentResolveMetadata getProcessedMetadata();
```

Comment:
⭕️ Alternatively, instead of returning `null`, this method could take a `Factory<ModuleComponentResolveMetadata>` that knows how to create it if missing.

Reply:
Yeah, I tried that and it didn't feel simpler or more elegant. A better solution would fix #4261 by caching per-project per-repository. But I just want to get a fix in.